crates/xtask/src/tmt.rs

```diff
@@ -215,6 +287,274 @@ fn parse_plan_metadata(plans_file: &Utf8Path) -> Result<std::collections::HashMa
+
+    if !test_dir.is_dir() {
+        // Legacy file-based test, no custom images
+        return Ok(Vec::new());
+    }
+
+    // Discover Containerfiles
+    let custom_images = discover_containerfiles(&test_dir, &test_name_prefix)?;
+
+    if custom_images.is_empty() {
+        return Ok(Vec::new());
+    }
+
+    println!(
+        "\nBuilding {} custom image(s) for {}...",
+        custom_images.len(),
+        plan_name
+    );
+
+    for image in &custom_images {
+        // Build from images/<tag>/ subdirectory
+        let image_dir = test_dir.join("images").join(&image.image_dir);
+        let containerfile_path = image_dir.join("Containerfile");
+        let image_ref = format!("{}/{}", registry_url, image.full_tag);
+
+        println!(
+            "  Building {} from {}/images/{}...",
+            image_ref, test_name_prefix, image.image_dir
+        );
+
+        // Build context is images/<tag>/ subdirectory
+        cmd!(
+            sh,
+            "podman build -t {image_ref} -f {containerfile_path} {image_dir}"
+        )
+        .run()
+        .with_context(|| {
+            format!(
+                "Failed to build {} from images/{}",
+                image.full_tag, image.image_dir
+            )
+        })?;
+
+        println!("  Pushing {} to registry...", image_ref);
+
+        // Push to registry
+        cmd!(sh, "skopeo copy --dest-tls-verify=false containers-storage:{image_ref} docker://{image_ref}")
+            .run()
+            .with_context(|| format!("Failed to push {}", image.full_tag))?;
+
+        println!("  Successfully pushed {}", image_ref);
+    }
+
+    Ok(custom_images)
+}
+
+/// Remove custom images for a test from local storage
+#[context("Cleaning up custom images")]
+fn cleanup_test_images(sh: &Shell, images: &[CustomImage]) -> Result<()> {
+    if images.is_empty() {
+        return Ok(());
+    }
+
+    println!("\nCleaning up {} custom image(s)...", images.len());
+
+    for image in images {
+        // Remove from local storage
+        // We don't remove from registry since it's a test registry that gets torn down anyway
+        let full_tag = &image.full_tag;
+        let _ = cmd!(sh, "podman rmi -f {full_tag}")
+            .ignore_stderr()
+            .ignore_status()
+            .run();
+
+        println!("  Removed local image {}", image.full_tag);
+    }
+
+    Ok(())
+}
+
+/// Information about a running registry VM
+struct RegistryVmInfo {
+    /// VM name
+    vm_name: String,
+    /// SSH port for VM access
+    ssh_port: u16,
+    /// SSH private key
+    ssh_key: String,
+    /// IP address where test VMs can reach the registry (host gateway IP)
+    registry_ip: String,
+}
+
+/// Setup and configure a persistent registry VM for test images
+/// Returns registry VM info and cleanup guard (which will cleanup VM on drop unless disarmed)
+#[context("Setting up registry VM")]
+fn setup_registry_vm<'a>(
```

Comment:
Hmm just so I understand here - is this only used for non-testing-farm cases i.e. in GHA? A toplevel goal of this effort is that with testing-farm we'd be spinning up two VMs in the IaaS it uses (e.g. AWS or whatever) right?